Review status: 0 of 2 files reviewed at latest revision, 12 unresolved discussions, some commit checks pending.

---

<sup>**[src/fancy-transformers/arithmetic.spec.ts, line 5 \[r1\]](https://reviewable.io:443/reviews/uproxy/uproxy-lib/294#-Jzqkz5lhCmIt9oOAGdX)** ([raw file](https://github.com/uproxy/uproxy-lib/blob/f1ea3f13cdc6db4787081ec9cd5791c794d8f40c/src/fancy-transformers/arithmetic.spec.ts#L5)):</sup>
These are already in alphabetical order.

---

<sup>**[src/fancy-transformers/arithmetic.spec.ts, line 10 \[r1\]](https://reviewable.io:443/reviews/uproxy/uproxy-lib/294#-Jzql4Jv40wctfSh3HDT)** ([raw file](https://github.com/uproxy/uproxy-lib/blob/f1ea3f13cdc6db4787081ec9cd5791c794d8f40c/src/fancy-transformers/arithmetic.spec.ts#L10)):</sup>
Okay I have made this change.

---

<sup>**[src/fancy-transformers/arithmetic.spec.ts, line 11 \[r1\]](https://reviewable.io:443/reviews/uproxy/uproxy-lib/294#-Jzqpn-aG-Hzy6_tS9H3)** ([raw file](https://github.com/uproxy/uproxy-lib/blob/f1ea3f13cdc6db4787081ec9cd5791c794d8f40c/src/fancy-transformers/arithmetic.spec.ts#L11)):</sup>
Okay I have made this change.

---

<sup>**[src/fancy-transformers/arithmetic.spec.ts, line 21 \[r1\]](https://reviewable.io:443/reviews/uproxy/uproxy-lib/294#-JzqlmFmfDav5NgD0YLS)** ([raw file](https://github.com/uproxy/uproxy-lib/blob/f1ea3f13cdc6db4787081ec9cd5791c794d8f40c/src/fancy-transformers/arithmetic.spec.ts#L21)):</sup>
Since the encoder and decoder objects are stateful, I think you need to make a new instance for every test to be safe.

---

<sup>**[src/fancy-transformers/arithmetic.spec.ts, line 36 \[r1\]](https://reviewable.io:443/reviews/uproxy/uproxy-lib/294#-Jzqm8Khdh1RPGRVBoBa)** ([raw file](https://github.com/uproxy/uproxy-lib/blob/f1ea3f13cdc6db4787081ec9cd5791c794d8f40c/src/fancy-transformers/arithmetic.spec.ts#L36)):</sup>
The encoding input is a real observed WebRTC packet that was breaking the encoder. The output is the encoding of that packet using a reference implementation. I have added comments to the test to say where the inputs and outputs come from.

---

<sup>**[src/fancy-transformers/arithmetic.ts, line 0 \[r1\]](https://reviewable.io:443/reviews/uproxy/uproxy-lib/294#-JzqmV3KHS1Ev8fMGc9D)** ([raw file](https://github.com/uproxy/uproxy-lib/blob/f1ea3f13cdc6db4787081ec9cd5791c794d8f40c/src/fancy-transformers/arithmetic.ts#L0)):</sup>
Okay I have made this change.

---

<sup>**[src/fancy-transformers/arithmetic.ts, line 0 \[r1\]](https://reviewable.io:443/reviews/uproxy/uproxy-lib/294#-JzqqLU8XGpCt3T8mOp6)** ([raw file](https://github.com/uproxy/uproxy-lib/blob/f1ea3f13cdc6db4787081ec9cd5791c794d8f40c/src/fancy-transformers/arithmetic.ts#L0)):</sup>
It looks like not: "arithmetic.ts(146,3): error TS1089: 'protected' modifier cannot appear on a constructor declaration."

---

<sup>**[src/fancy-transformers/arithmetic.ts, line 0 \[r1\]](https://reviewable.io:443/reviews/uproxy/uproxy-lib/294#-Jzqmebqrrj4uENN3RfZ)** ([raw file](https://github.com/uproxy/uproxy-lib/blob/f1ea3f13cdc6db4787081ec9cd5791c794d8f40c/src/fancy-transformers/arithmetic.ts#L0)):</sup>
I rewrote this using map.

---

<sup>**[src/fancy-transformers/arithmetic.ts, line 0 \[r1\]](https://reviewable.io:443/reviews/uproxy/uproxy-lib/294#-Jzqq6pMHDQrwpXe4oH9)** ([raw file](https://github.com/uproxy/uproxy-lib/blob/f1ea3f13cdc6db4787081ec9cd5791c794d8f40c/src/fancy-transformers/arithmetic.ts#L0)):</sup>
I rewrote this using reduce.

---

<sup>**[src/fancy-transformers/arithmetic.ts, line 0 \[r1\]](https://reviewable.io:443/reviews/uproxy/uproxy-lib/294#-Jzqq-RsHIibsjbyFdUR)** ([raw file](https://github.com/uproxy/uproxy-lib/blob/f1ea3f13cdc6db4787081ec9cd5791c794d8f40c/src/fancy-transformers/arithmetic.ts#L0)):</sup>
There is no specific reference implementation for the code. There was psuedocode in http://www.arturocampos.com/ac_range.html that gave a general. However, it was incomplete and I filled in the gaps by looking at the other papers. 

---

<sup>**[src/fancy-transformers/arithmetic.ts, line 0 \[r1\]](https://reviewable.io:443/reviews/uproxy/uproxy-lib/294#-JzqmvnlB-LE2ht5-iZ3)** ([raw file](https://github.com/uproxy/uproxy-lib/blob/f1ea3f13cdc6db4787081ec9cd5791c794d8f40c/src/fancy-transformers/arithmetic.ts#L0)):</sup>
Okay I have made this change.

---


Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/uproxy/uproxy-lib/294)
<!-- Sent from Reviewable.io -->

--- a/src/fancy-transformers/arithmetic.spec.ts
+++ b/src/fancy-transformers/arithmetic.spec.ts
@@ -14,13 +14,20 @@ function makeUniformProbabilities() :number[] {
   return probs;
 }
 
+let encoder :arithmetic.Encoder;
+let decoder :arithmetic.Decoder;
+
 describe('Arithmetic coding and decoding - short inputs', function() {
+  beforeEach(function() {
+    encoder = new arithmetic.Encoder(makeUniformProbabilities());
+    decoder = new arithmetic.Decoder(makeUniformProbabilities());
+  });
+
   // The encoding input is a simple numerical sequence.
   // The encoding output is taken from a reference implementation.
   it('encoding short input', function() {
     let plain = arraybuffers.hexStringToArrayBuffer("0.1.2.3");
     let target = arraybuffers.hexStringToArrayBuffer("ca.0.1.2.3.0.0.0.8");
-    let encoder = new arithmetic.Encoder(makeUniformProbabilities());
     let result = encoder.encode(plain);
     expect(arraybuffers.byteEquality(result, target)).toBe(true);
   });
@@ -29,13 +36,17 @@ describe('Arithmetic coding and decoding - short inputs', function() {
   it('decoding short input', function() {
     let encoded = arraybuffers.hexStringToArrayBuffer("ca.0.1.2.3.0.0.0.8");
     let target = arraybuffers.hexStringToArrayBuffer("0.1.2.3");
-    let decoder = new arithmetic.Decoder(makeUniformProbabilities());
     let result = decoder.decode(encoded);
     expect(arraybuffers.byteEquality(result, target)).toBe(true);
   });
 });
 
 describe('Arithmetic coding and decoding - long inputs', function() {
+  beforeEach(function() {
+    encoder = new arithmetic.Encoder(makeUniformProbabilities());
+    decoder = new arithmetic.Decoder(makeUniformProbabilities());
+  });
+
   // The encoding input is an example of a real WebRTC packet.
   // The encoding output is taken from a reference implementation.
   it('encoding long input', function() {

